mbtools/replace_im_links.py

```diff
@@ -0,0 +1,71 @@
+import json
+from mbtools import utils
+import argparse
+from lxml import html
+from pathlib import Path
+
+IM_PREFIX = "https://s3.amazonaws.com/im-ims-export/"
+# Make Conditional for local instance
+CONTENT_PREFIX = "https://k12.openstax.org/contents/raise/"
+
+
+def parse_media_file(media_path, s3_prefix):
+    mapping = {}
+    with open(media_path) as f:
+        data = json.load(f)
+        for entry in data:
+            im_url = IM_PREFIX + entry['original_filename']
+            s3_url = s3_prefix + '/' + entry['s3_key']
+            mapping[im_url] = s3_url
+    return mapping
+
+
+def replace_im_links(content_path, media_path, s3_prefix, mode):
+    im_to_osx_mapping = parse_media_file(media_path, s3_prefix)
+    replacements = {}
+    if (mode == 'mbz'):
+        backup = utils.parse_moodle_backup(content_path)
+        activities = backup.activities()
+        q_bank = backup.q_bank
+        for act in activities:
+            elems = act.html_elements()
+            for elem in elems:
+                changes = elem.replace_attribute_values(
+                              ['src', 'href'], im_to_osx_mapping)
+                if (len(changes) > 0):
+                    replacements.update(changes)
+                    utils.write_etree(act.activity_filename, act.etree)
+                    utils.write_etree(q_bank.questionbank_path, q_bank.etree)
+        return replacements
+    elif (mode == 'html'):
+        for item in content_path.iterdir():
+            content_tree = None
+            with open(item, 'r') as f:
+                data = f.read()
+                content_tree = html.fragments_fromstring(data)[0]
+                utils.replace_attribute_values_tree(
+                    content_tree, ['src', 'href'], im_to_osx_mapping
+                    )
+            with open(item, 'w') as f:
+                f.write(html.tostring(content_tree).decode())
```

Comment:
I'm fairly certain that serializing things with `html.tostring()` is going to have all of the issues that led me to write [this method](https://github.com/openstax/raise-mbtools/blob/6f9752a5815410486f22d3da470d86debf13cbc2/mbtools/models.py#L253-L260) the way that I did. Perhaps what we should do is define a function in `models.py` like the following:

```python
def html_fragments_to_string(fragments):
    # Pass things through bs4 so we can avoid adding closing tags and
    # closing slashes that lxml may otherwise emit on void elements
    html_content = b''.join(
        [etree.tostring(fragment) for fragment in fragments])
    return BeautifulSoup(
        html_content,
        "html.parser"
    ).encode(formatter="html5").decode('utf-8')
```

Then `MoodleHtmlElement.tostring()` becomes:

```python
    def tostring(self):
        return html_fragments_to_string(self.etree_fragments)
```

Finally the code here maybe can be something like:

```python
                data = f.read()
                fragments = html.fragments_fromstring(data)
                utils.replace_attribute_values_tree(
                    fragments[0], ['src', 'href'], im_to_osx_mapping
                    )
            with open(item, 'w') as f:
                f.write(models.html_fragments_to_string(fragments))
```